### PR TITLE
Update the Jupyter UI for 0.5

### LINKED ISSUE #2588: jupyter web app should provide status of notebook
Here's a screen shot of the Jupyter spawner after creating a new notebook

My notebook shows up but there is no indication of the status (e.g. pending, starting, running) etc...

We should expose that info in the UI

jupyterspawner

### LINKED ISSUE #2592: notebook manager - Volume name should default to name of notebook
Right now the volume used for the notebook home directory /home/jovyan always defaults to "notebook-workspace".

This default will be invalid if the user tries to create multiple notebooks in a given namespace because the volume would already exist.

I think it would be better to make the volume name default to the name of the notebook so it would default to being unique for each notebook

/cc @vkoukis @kimwnasptd

### LINKED ISSUE #2639: Notebook manager should provide a UI for visualizing jupyter spec
Once a notebook is launched, it would be useful to be able to visualize the jupyter spec (e.g. attached volumes, image, resources, etc...).

Right now the only options in the UI are
* Connect
* Delete

Perhaps some of these fields could be added as columns in the table listing running notebooks?

Should we try to get this done in 0.5? Its not critical but on the other hand it improves the usability / fit & finish of the new notebook story which is a major part of the 0.5 release. 

/cc @vkoukis @kimwnasptd

### LINKED ISSUE #2834: Make sure error messages appear where the user expects them in the Jupyter spawner UI
Steps to reproduce

1. Go to the notebook manager UI and click "+" to spawn a notebook
1. Fill out the form with an invalid name e.g. a notebook name with spaces
1. Click spawn

An error in this case pops up but it shows up at the top of the page. See screenshot below.

The problem is if the top of the page is above the fold (e.g. user is on a laptop) then the error isn't obvious to the user. it just looks like the user clicked spawn and nothing happened.

error

/cc @avdaredevil @kimwnasptd @vkoukis

### PR DESCRIPTION
**Pull request by kimwnasptd**
closes #2639
closes #2592
closes #2834
closes #2588

alt text

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/2848)

### CODE REVIEW

File: components/jupyter-web-app/kubeflow_jupyter/rok/static/js/notebooks.js [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+  var ns = new URL(window.location.href).searchParams.get("namespace")
+  if (ns) {
+    // Load the selected ns
+    $("#ns-select").val(ns)
+  }
+  $("#ns-select").trigger("change");
+
+  // Periodically check if the Notebook's State has changed
+  if (rok_token.length > 0) {
+    var inrvl = setInterval(function() {
```
[the remark this answers is not in the document]
Comment (kimwnasptd) [reply components/jupyter-web-app/kubeflow_jupyter/rok/static/js/notebooks.js]: Changed the polling mechanism to exponential backoff. 
The polling will happen in batches of 3 requests. In the first batch the requests will be issued each second. If the state doesn't change in these 3, then in the next batch the requests will be sent each 2 seconds etc. The cap will be 16 seconds between requests.

If the state changes in one of these requests, or the user deletes/creates a notebook or selects a different namespace; the polling mechanism will reset and the requests will fire again each second.

WDYT?

**Comment by prodonjs** [comment]
This looks good. Definitely looks more native inside of the parent iframe and the exponential backoff seems like a good approach. It does seem unfortunate that the code is duplicated inside the `default` and `rok` directories. I don't have any context for why those exist separately, but it does seem like a maintenance burden to have to replicate changes to both. Perhaps that's something to flag as a future improvement.

In any case, this looks good given my limited scope of assessment.
/lgtm
/approve

**Comment by jlewi** [comment]
@kimwnasptd Is this PR still WIP? If not can we remove WIP from the title?

**Comment by abhi-g** [comment]
Looks like the presubmit test failed trying to cleanup after itself. Retrying once more.

/retest kubeflow-presubmit

**Pull request merged**